src/promptflow-tools/promptflow/tools/common.py

```diff
@@ -56,15 +61,33 @@ def __str__(self):
         return "\n".join(map(str, self))
 
 
-def validate_role(role: str, valid_roles: List[str] = None):
+class PromptResult(str):
+    def __init__(self, string):
+        super().__init__()
+        self.original_string = string
+        self.escaped_string = ""
+        self.escaped_mapping = {}
+
+    def get_escape(self) -> str:
+        return self.escaped_string
+
+    def get_escape_mapping(self) -> dict:
+        return self.escaped_mapping
+
```

Comment:
may add methods to set escaped_string/mapping, which could make logic more clear.